### ISSUE TITLE
Panic trying to unwrap a rect that could not be transformed

### ISSUE THREAD

**Issue by nical** [servo/webrender#2911]
Filed as a followup of #2904.

The panic is because this https://github.com/servo/webrender/pull/2904/files#diff-9c1d8b23da9e37ae3377086804c5aaf2L588 used to always have Some(transformed rect) and now can have None since transforming a rect is fallible in the latest euclid, since servo/euclid#291.

cc @kvark

I suppose that in this case we could cull out entirely the rects that we can't project.

This was caught by Firefox's CI, see https://bugzilla.mozilla.org/show_bug.cgi?id=1476636#c1
Here's a sample log: https://treeherder.mozilla.org/logviewer.html#?job_id=188772693&repo=try&lineNumber=1438

**Comment by kvark** [comment]
@nical removing `Some` in any place is not causing a panic, `unwrap` is. Which unwrap? what is the call stack?

**Comment by nical** [comment]
The call stack is partly visible in the sample log, although some inlining probably got in the way of showing some of the interesting bits.
We know that we are inside of `prepare_prim_runs`. My best guess is that we are crashing in [one of the three calls](https://github.com/Eijebong/webrender/blob/290335727583f922b78714305ed6c5abf4842840/webrender/src/prim_store.rs#L2759) to `FastTransfrom::transform_rect` which forwards to `Transform3D::transform_rect` [and unwraps the result](https://github.com/Eijebong/webrender/blob/290335727583f922b78714305ed6c5abf4842840/webrender/src/util.rs#L579).
Before the euclid update this always returned a rect even if the projection didn't make sense, and now in some cases we a `None` that we need to deal with somehow.

Yesterday I merely gave it a quick look before going home so I don't remember how I linked things to this https://github.com/servo/webrender/pull/2904/files#diff-9c1d8b23da9e37ae3377086804c5aaf2L588 previously always having Some(transformed rect), but in any case this also a previously infallible code path that can now get `None` and we should make sure to review our past assumptions about that as well as the other code paths indirectly affected by the result of `transform_rect`.

**Comment by staktrace** [comment]
For the record, I reproduced this crash on a local build from [this revision](https://treeherder.mozilla.org/#/jobs?repo=try&revision=bb136e39ff30998ee001d6e390efe9fe674a3546):

STR: `MOZ_WEBRENDER=1 MOZ_ACCELERATED=1 ./mach crashtest dom/animation/test/crashtests/1272475-1.html`

Some relevant stack frames:
```
#8  0x00007fffe94728b3 in core::panicking::panic_fmt::h61b193aee9efb823 () at libcore/panicking.rs:71
#9  0x00007fffe94727cb in core::panicking::panic::heb63751038cba1dc () at libcore/panicking.rs:51
#10 0x00007fffe8c47b1d in _$LT$core..option..Option$LT$T$GT$$GT$::unwrap::h75a81ddd1418cd25 (self=...) at /checkout/src/libcore/macros.rs:20
#11 0x00007fffe8c5144f in webrender::prim_store::PrimitiveStore::prepare_prim_runs::h1747d903823c2d31 (rect=0x0, self=<optimized out>) at gfx/webrender/src/util.rs:579
#12 0x00007fffe8c5144f in webrender::prim_store::PrimitiveStore::prepare_prim_runs::h1747d903823c2d31 (self=<optimized out>, pic_context=0x7fffc14f6b90, pic_state=0x7fffc14f6cd0, frame_context=<optimized out>, frame_state=0x7fffc14f8c40) at gfx/webrender/src/prim_store.rs:2758
#13 0x00007fffe8c506c0 in webrender::prim_store::PrimitiveStore::prepare_prim_for_render::hfc654b7fc5c94236 (self=<optimized out>, prim_index=..., prim_run_context=<optimized out>, pic_context=0x7fffc14f7240, pic_state=0x7fffc14f7380, frame_context=0x7fffc14f8c90, frame_state=0x7fffc14f8c40) at gfx/webrender/src/prim_store.rs:2505
#14 0x00007fffe8c5137d in webrender::prim_store::PrimitiveStore::prepare_prim_runs::h1747d903823c2d31 (self=<optimized out>, pic_context=0x7fffc14f7240, pic_state=0x7fffc14f7380, frame_context=<optimized out>, frame_state=0x7fffc14f8c40) at gfx/webrender/src/prim_store.rs:2739
#15 0x00007fffe8c506c0 in webrender::prim_store::PrimitiveStore::prepare_prim_for_render::hfc654b7fc5c94236 (self=<optimized out>, prim_index=..., prim_run_context=<optimized out>, pic_context=0x7fffc14f78f0, pic_state=0x7fffc14f7a30, frame_context=0x7fffc14f8c90, frame_state=0x7fffc14f8c40) at gfx/webrender/src/prim_store.rs:2505
#16 0x00007fffe8c5137d in webrender::prim_store::PrimitiveStore::prepare_prim_runs::h1747d903823c2d31 (self=<optimized out>, pic_context=0x7fffc14f78f0, pic_state=0x7fffc14f7a30, frame_context=<optimized out>, frame_state=0x7fffc14f8c40) at gfx/webrender/src/prim_store.rs:2739
```

The unwrap in question is [the one](https://github.com/servo/webrender/blob/1ca448a162ef2aee8f6ed232200d382719fb405c/webrender/src/util.rs#L579) in `transform_rect` in util.rs.

**Comment by kvark** [comment]
I believe this is fixed by #2913 plus #2947
